### PR TITLE
Cleanup error handling of cbmc/ folder

### PR DESCRIPTION
**Pull request by NlightNFotis**
Perform various cleanups of the error handling mechanisms in the `src/cbmc/` folder. These include turning `assert`s into `INVARIANT`s and `PRECONDITION`s as well as a new custom exception class.

### CODE REVIEW

File: src/cbmc/symex_coverage.cpp [hunk 1]
```diff
@@ -139,15 +139,17 @@ goto_program_coverage_recordt::goto_program_coverage_recordt(
...
 {
-  assert(gf_it->second.body_available());
+  PRECONDITION(gf_it->second.body_available());
 
   // identify the file name, inlined functions aren't properly
   // accounted for
   goto_programt::const_targett end_function=
     --gf_it->second.body.instructions.end();
-  assert(end_function->is_end_function());
+  INVARIANT(end_function->is_end_function(),
```
Comment (martin-cs) [inline src/cbmc/symex_coverage.cpp]: DATA_INVARIANT I think.

File: src/cbmc/symex_coverage.cpp [hunk 2]
```diff
@@ -139,15 +139,17 @@ goto_program_coverage_recordt::goto_program_coverage_recordt(
...
+    "Last instruction in a function body is end function");
   file_name=end_function->source_location.get_file();
-  assert(!file_name.empty());
+  INVARIANT(!file_name.empty(),
```
Comment (martin-cs) [inline src/cbmc/symex_coverage.cpp]: DATA_INVARIANT

File: src/cbmc/symex_coverage.cpp [hunk 3]
```diff
@@ -260,11 +262,14 @@ void goto_program_coverage_recordt::compute_coverage_lines(
         for(const auto &cov : c_entry->second)
           std::cerr << cov.second.succ->location_number << '\n';
       }
-      assert(c_entry->second.size()==1 || is_branch);
+      INVARIANT(c_entry->second.size()==1 || is_branch,
+        "Instructions other than branch instructions have exactly 1 successor");
```
Comment (martin-cs) [inline src/cbmc/symex_coverage.cpp]: DATA_INVARIANT

File: src/util/parse_options.cpp [hunk 4]
```diff
@@ -82,3 +84,19 @@ banner_string(const std::string &front_end, const std::string &version)
...
+  // catch all exceptions here so that this code is not duplicated
+  // for each tool
+  try
+  {
+    return parse_options_baset::main();
+  }
+  catch(invalid_user_input_exceptiont &e)
+  {
+    std::cerr << e.what() << "\n";
+    return CPROVER_EXIT_EXCEPTION;
```
Comment (martin-cs) [inline src/util/parse_options.cpp]: CPROVER_EXIT_USAGE_ERROR ?

**Review by martin-cs — APPROVED** [review]
Approve subject to addressing the comments.

File: src/cbmc/bmc_cover.cpp [hunk 5]
```diff
@@ -241,7 +241,8 @@ bool bmc_covert::operator()()
     cover_goals.add(l);
   }
 
-  assert(cover_goals.size()==goal_map.size());
+  INVARIANT(cover_goals.size()==goal_map.size(),
```
Comment (peterschrammel) [inline src/cbmc/bmc_cover.cpp]: apply clang-format

File: src/cbmc/cbmc_main.cpp [hunk 6]
```diff
@@ -41,9 +41,11 @@ int wmain(int argc, const wchar_t **argv_wide)
...
-  cbmc_parse_optionst parse_options(argc, argv);
+  try
+  {
+    cbmc_parse_optionst parse_options(argc, argv);
 
-  int res=parse_options.main();
+    int res = parse_options.main();
 
   #ifdef IREP_HASH_STATS
   std::cout << "IREP_HASH_CNT=" << irep_hash_cnt << '\n';
```
Comment (peterschrammel) [inline src/cbmc/cbmc_main.cpp]: indentation

File: src/util/exception_utils.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    std::string option,
+    std::string correct_input = "")
+    : reason(reason), option(option), correct_input(correct_input)
+  {
+  }
+
+  std::string what() const noexcept
+  {
+    std::string res;
+    res += "\nInvalid User Input Exception\n";
```
Comment (peterschrammel) [inline src/util/exception_utils.h]: I'd remove the word `Exception`.

File: src/cbmc/cbmc_solvers.cpp [hunk 8]
```diff
@@ -235,20 +239,38 @@ std::unique_ptr<cbmc_solverst::solvert> cbmc_solverst::get_smt2(
...
 {
-  if(options.get_bool_option("beautify"))
+  bool beautify = options.get_bool_option("beautify");
+
+  if(beautify)
   {
-    error() << "sorry, this solver does not support beautification" << eom;
-    throw 0;
+    throw invalid_user_input_exceptiont(
+      "Sorry, this solver does not support beautification",
```
Comment (peterschrammel) [inline src/cbmc/cbmc_solvers.cpp]: maybe `this solver` -> `the chosen solver` ? (same below)

File: src/util/parse_options.cpp [hunk 9]
```diff
@@ -82,3 +84,19 @@ banner_string(const std::string &front_end, const std::string &version)
...
+{
+  // catch all exceptions here so that this code is not duplicated
+  // for each tool
+  try
+  {
+    return parse_options_baset::main();
+  }
+  catch(invalid_user_input_exceptiont &e)
+  {
+    std::cerr << e.what() << "\n";
```
Comment (peterschrammel) [inline src/util/parse_options.cpp]: No. This must go into a message stream.

File: src/cbmc/cbmc_main.cpp [hunk 10]
```diff
@@ -41,11 +41,9 @@ int wmain(int argc, const wchar_t **argv_wide)
 int main(int argc, const char **argv)
 {
 #endif
-  try
```
Comment (peterschrammel) [inline src/cbmc/cbmc_main.cpp]: Please squash this with the first commit that introduce the change that you undo here.

**Review by peterschrammel — CHANGES_REQUESTED** [review]

File: src/cbmc/bmc_cover.cpp [hunk 11]
```diff
@@ -241,7 +241,8 @@ bool bmc_covert::operator()()
...
+    "We add coverage for each goal");
```
Comment (tautschnig) [inline src/cbmc/bmc_cover.cpp]: The coding standard says that error messages start with a lower-case letter.

File: src/cbmc/symex_coverage.cpp [hunk 12]
```diff
@@ -260,11 +262,14 @@ void goto_program_coverage_recordt::compute_coverage_lines(
...
 
       for(const auto &cov : c_entry->second)
       {
-        assert(cov.second.num_executions>0);
+        INVARIANT(cov.second.num_executions>0,
+          // FIXME I don't know what this means
+          "Number of executions most be positive");
```
Comment (tautschnig) [inline src/cbmc/symex_coverage.cpp]: Should be a `PRECONDITION`. A coverage entry can only exist when a location has actually been covered in at least one execution.

File: src/cbmc/cbmc_main.cpp [hunk 13]
```diff
@@ -52,4 +54,9 @@ int main(int argc, const char **argv)
   #endif
 
   return res;
+  }
+  catch(invalid_user_input_exceptiont &e)
+  {
+    std::cerr << e.what() << std::endl;
+  }
```
Comment (tautschnig) [inline src/cbmc/cbmc_main.cpp]: Design note: I don't think such an exception should propagate out of `parse_options` for the key task of `parse_options` is to parse options? If anything goes wrong it should and does return an error code. Having both error codes and exceptions seems wrong to me.

File: src/util/exception_utils.h [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+public:
+  invalid_user_input_exceptiont(
+    std::string reason,
+    std::string option,
+    std::string correct_input = "")
+    : reason(reason), option(option), correct_input(correct_input)
+  {
+  }
+
+  std::string what() const noexcept
```
Comment (tautschnig) [inline src/util/exception_utils.h]: I don't think this should be a header-only class. Add a cpp file and move the implementation of at least this function over there. No need to inline it everywhere, will just unnecessarily increase the size of the binary.

File: src/cbmc/bmc.cpp [hunk 15]
```diff
@@ -14,6 +14,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <chrono>
 #include <iostream>
 
+#include <util/exception_utils.h>
```
Comment (tautschnig) [inline src/cbmc/bmc.cpp:17]: Random note: I don't understand the commit message. What is an "unstructured commit?"

File: src/cbmc/bmc.cpp [hunk 16]
```diff
@@ -296,9 +297,8 @@ void bmct::get_memory_model()
     memory_model=util_make_unique<memory_model_psot>(ns);
   else
   {
-    error() << "Invalid memory model " << mm
-            << " -- use one of sc, tso, pso" << eom;
-    throw "invalid memory model";
+    throw invalid_user_input_exceptiont(
+      "Invalid parameter " + mm, "--mm", "values of sc, tso, pso");
```
Comment (tautschnig) [inline src/cbmc/bmc.cpp]: I'm struggling to piece this together: what sort of sentence is going to be formed using "values of ..."?

File: src/cbmc/bv_cbmc.cpp [hunk 17]
```diff
@@ -27,12 +24,8 @@ bvt bv_cbmct::convert_waitfor(const exprt &expr)
...
 
   mp_integer bound_value;
-  if(to_integer(bound, bound_value))
-  {
-    error().source_location=expr.find_source_location();
-    error() << "waitfor bound must be a constant" << eom;
-    throw 0;
-  }
+  bool successful_cast = to_integer(bound, bound_value);
+  INVARIANT(successful_cast, "waitfor bound must be a constant");
```
Comment (tautschnig) [inline src/cbmc/bv_cbmc.cpp]: Just use `numeric_cast`?

File: src/cbmc/cbmc_solvers.cpp [hunk 18]
```diff
@@ -169,8 +171,10 @@ std::unique_ptr<cbmc_solverst::solvert> cbmc_solverst::get_smt2(
   {
     if(solver==smt2_dect::solvert::GENERIC)
     {
-      error() << "please use --outfile" << eom;
-      throw 0;
+      throw invalid_user_input_exceptiont(
+        "Failed to provide a filename",
```
Comment (tautschnig) [inline src/cbmc/cbmc_solvers.cpp]: Who failed?

File: src/cbmc/cbmc_solvers.cpp [hunk 19]
```diff
@@ -212,8 +216,8 @@ std::unique_ptr<cbmc_solverst::solvert> cbmc_solverst::get_smt2(
 
     if(!*out)
     {
-      error() << "failed to open " << filename << eom;
-      throw 0;
+      throw invalid_user_input_exceptiont(
+        "Failed to open suggested file: " + filename, "--outfile");
```
Comment (tautschnig) [inline src/cbmc/cbmc_solvers.cpp]: Well, when I work with a deterministic computer program I'm not making suggestions.

File: src/cbmc/cbmc_solvers.cpp [hunk 20]
```diff
@@ -235,20 +239,38 @@ std::unique_ptr<cbmc_solverst::solvert> cbmc_solverst::get_smt2(
 
 void cbmc_solverst::no_beautification()
 {
-  if(options.get_bool_option("beautify"))
+  bool beautify = options.get_bool_option("beautify");
```
Comment (tautschnig) [inline src/cbmc/cbmc_solvers.cpp]: Why is this temporary introduced?

File: src/cbmc/cbmc_solvers.cpp (hunk 8 above)
Comment (tautschnig) [reply src/cbmc/cbmc_solvers.cpp]: And drop "sorry" - we don't apologise elsewhere either.

File: src/cbmc/cbmc_solvers.cpp [hunk 21]
```diff
@@ -235,20 +239,38 @@ std::unique_ptr<cbmc_solverst::solvert> cbmc_solverst::get_smt2(
...
+      "--beautify");
   }
 }
 
 void cbmc_solverst::no_incremental_check()
 {
-  if(options.get_bool_option("all-properties") ||
-     options.get_option("cover")!="" ||
-     options.get_option("incremental-check")!="")
+  bool all_properties = options.get_bool_option("all-properties");
```
Comment (tautschnig) [inline src/cbmc/cbmc_solvers.cpp]: `const` (here and below)

File: src/cbmc/cbmc_solvers.cpp [hunk 22]
```diff
@@ -235,20 +239,38 @@ std::unique_ptr<cbmc_solverst::solvert> cbmc_solverst::get_smt2(
...
+  bool cover = options.get_option("cover") != "";
```
Comment (tautschnig) [inline src/cbmc/cbmc_solvers.cpp]: While you're cleaning this up: use `.is_set` instead of `get_option(...) != ""`

File: src/util/parse_options.h [hunk 23]
```diff
@@ -35,6 +35,17 @@ class parse_options_baset
   bool parse_result;
 };
 
+class parse_optionst : public parse_options_baset
```
Comment (tautschnig) [inline src/util/parse_options.h]: I'm not sure we need this derived class: can't it just be changed in `parse_options_baset` and make everyone benefit without changes?

**Review by tautschnig — COMMENTED** [review]
This looks like a good direction of work to me, but I'd suggest to make this as clean as possible unless there is a need to rush it in? If so, please break it apart into separate pull requests as the need for further discussion varies widely among the commits.

File: src/cbmc/symex_coverage.cpp (hunk 12 above)
Comment (NlightNFotis) [reply src/cbmc/symex_coverage.cpp]: There's a problem with that, in that the `PRECONDITION` according to documentation should only contain unmodified arguments to a function, but I can change it, since it makes it easier to drop the message which appears incorrect as well.

Comment (tautschnig) [reply src/cbmc/symex_coverage.cpp]: You might use a `DATA_INVARIANT`, paraphrasing in the message what I said above (coverage entries can only exist with at least one execution).

File: src/cbmc/bmc.cpp (hunk 15 above)
Comment (NlightNFotis) [reply src/cbmc/bmc.cpp:17]: It should read "unstructured assert", I have no idea what happened with that. Might have messed up with a rewording of a commit. Good catch, thanks.

File: src/cbmc/bmc.cpp (hunk 16 above)
Comment (NlightNFotis) [reply src/cbmc/bmc.cpp]: These all are parameters of the new class' constructor. In the end, `what()` is supposed to show something like this:
```
Invalid User Input
Option: -mm
Reason: Invalid parameter <placeholder>
Try values of sc, tso, pso
```

File: src/cbmc/cbmc_solvers.cpp (hunk 18 above)
Comment (NlightNFotis) [reply src/cbmc/cbmc_solvers.cpp]: The user, but this is implicit in the error message showed by the `what()` function of this exception, which will look like this:

```
Invalid User Input
Option: --outfile
Reason: Failed to provide a filename
```

Comment (NlightNFotis) [reply src/cbmc/cbmc_solvers.cpp]: But I guess there's also an issue with the error string, it should start with a lowercase letter.

File: src/cbmc/cbmc_solvers.cpp (hunk 20 above)
Comment (NlightNFotis) [reply src/cbmc/cbmc_solvers.cpp]: I remember I had a reason in the beginning but I may have refactored it enough that it's now unneeded.

File: src/cbmc/cbmc_main.cpp (hunk 13 above)
Comment (NlightNFotis) [reply src/cbmc/cbmc_main.cpp]: This is a comment on outdated code. The error handling is now being contained in `parse_optionst::main()`, and doesn't escape that context. `parse_optionst::main()` will indeed return an error code, that is is being handled in `cbmc_main`

File: src/cbmc/bmc.cpp (hunk 16 above)
Comment (tautschnig) [reply src/cbmc/bmc.cpp]: Yes, but in order to figure out what to put in those arguments I will almost always have to look at the code generating the string. I don't think it's particularly obvious what (not) to put in the arguments. It would already be an improvement if the "Try " were not automatically included so that the argument is a full sentence?

File: src/cbmc/cbmc_solvers.cpp (hunk 18 above)
Comment (tautschnig) [reply src/cbmc/cbmc_solvers.cpp]: No, users never fail, by definition. You can ask them to do something different ("Please provide a filename"), or you can firmly order them to do so ("Filename required"/"Required filename not provided").

File: src/cbmc/cbmc_main.cpp (hunk 13 above)
Comment (NlightNFotis) [reply src/cbmc/cbmc_main.cpp]: Btw, it also appears that `parse_optionst` was mainly useful during the evolution of the code, but now that has been rebased, its usefulness has diminished, so I am trying to eliminate it.

File: src/util/parse_options.cpp (hunk 9 above)
Comment (NlightNFotis) [reply src/util/parse_options.cpp]: It was like this in the original design, but then I saw during rebasing on develop, that recently a similar change was eliminated, so I thought that it must have been that this is more appropriate. If I am also to do that now, I will need to add a `messaget` into `parse_options_baset`, which will require changes in various places

File: src/util/parse_options.h (hunk 23 above)
Comment (NlightNFotis) [reply src/util/parse_options.h]: This is not needed anymore. I have eliminated it.

File: src/cbmc/cbmc_solvers.cpp (hunk 18 above)
Comment (NlightNFotis) [reply src/cbmc/cbmc_solvers.cpp]: I think I will go with the last string, looks more reasonable in this context in my opinion `"Required filename not provided"`

File: src/util/exception_utils.h [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+  {
+    std::string res;
+    res += "\nInvalid User Input Exception\n";
+    res += "Option: " + option + "\n";
+    res += "Reason: " + reason;
+    // Print an optional correct usage message assuming correct input parameters have been passed
+    res += correct_input.empty() ? "\n" : " Try: " + correct_input + "\n";
```
Comment (danpoe) [inline src/util/exception_utils.h]: The `"Try: ..."` portion should better be on a new line.

Comment (NlightNFotis) [reply src/util/exception_utils.h]: @tautschnig Suggested to drop the `Try` altogether, and just add in the various constructor callsites so that it reads more like a sentence.

File: src/util/parse_options.cpp (hunk 9 above)
Comment (NlightNFotis) [reply src/util/parse_options.cpp]: I found exactly what I was referring to. So `parse_options_baset` doesn't seem to inherit from `messaget` and doesn't contain a reference to a `messaget`. If I do this, then I get loads of compilation errors regarding ambiguous calls to things like `status()`, `get_message_handler()`, etc.

It's also consistent with the rest of the file. `parse_options_baset::usage_error()`, `parse_options_baset::unknown_options_msg()` seem to be following the same pattern of using `std::cerr`

Comment (tautschnig) [reply src/util/parse_options.cpp]: At the risk of contradicting @peterschrammel: I don't think a message stream can be used here, because setting up the desired one requires successful parsing of command-line options (picking, e.g., JSON output with a verbosity of 5). Hence I'm with @NlightNFotis that using `std::cerr` is correct here.

**Comment by NlightNFotis** [comment]
@peterschrammel You have a red (blocker) cross applied to this. All your points have now been addressed. Can you take a look now and tell me if there's anything more you would want explained or changed?

**Comment by NlightNFotis** [comment]
@tautschnig Hi, can you also take a look to see if it's closer to what you expect? It's the same as before, but cleaned up and the changes suggested applied. The only extra is the last commit, which I thought makes sense to not be squashed, as it is a good atomic change (it replaces `to_integer` with `numeric_cast`).

File: src/cbmc/cbmc_parse_options.h [hunk 25]
```diff
@@ -76,10 +76,9 @@ class optionst;
   "(claim):(show-claims)(floatbv)(all-claims)(all-properties)" // legacy, and will eventually disappear // NOLINT(whitespace/line_length)
 // clang-format on
 
-class cbmc_parse_optionst:
-  public parse_options_baset,
-  public xml_interfacet,
-  public messaget
+class cbmc_parse_optionst : public parse_options_baset,
```
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.h]: what was the change here?

File: src/cbmc/cbmc_main.cpp [hunk 26]
```diff
@@ -41,9 +41,11 @@ int wmain(int argc, const wchar_t **argv_wide)
 int main(int argc, const char **argv)
 {
 #endif
-  cbmc_parse_optionst parse_options(argc, argv);
+  try
+  {
```
Comment (peterschrammel) [inline src/cbmc/cbmc_main.cpp]: This change is undone two commits later. Why is this change made at all?

File: src/util/exception_utils.cpp [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+
+Module: Exception helper utilities
+
+Author: Fotis Koutoulakis, fotis.koutoulakis@diffblue.com
+
+\*******************************************************************/
+
+#include "exception_utils.h"
+
+std::string invalid_user_input_exceptiont::what() const noexcept
```
Comment (peterschrammel) [inline src/util/exception_utils.cpp:11]: Ideally, there should also be a version hat can serialise into json and XML in a more structure way (in a future PR).

File: src/cbmc/bv_cbmc.cpp [hunk 28]
```diff
@@ -9,16 +9,14 @@ Author: Daniel Kroening, kroening@kroening.com
...
 {
-  if(expr.operands().size()!=4)
-  {
-    error().source_location=expr.find_source_location();
-    error() << "waitfor expected to have four operands" << eom;
-    throw 0;
-  }
+  DATA_INVARIANT(
+    expr.operands().size() != 4,
+    "waitfor expression is expected to have four operands");
```
Comment (peterschrammel) [inline src/cbmc/bv_cbmc.cpp]: @chrisr-diffblue, @martin-cs, the source location is lost here. Would it be possible to have variants of the INVARIANT macros that take a source location explicitly so that it can be output in a structured way?

**Review by peterschrammel — CHANGES_REQUESTED** [review]

**Comment by peterschrammel** [comment]
@NlightNFotis, the changes in f257667 that are later undone in 017c1b0 need to be removed.

**Comment by tautschnig** [comment]
Could this one please be rebased and cleaned up with regard to the comments raised by @peterschrammel - then it might be good to go.

**Comment by NlightNFotis** [comment]
@tautschnig Yes, this is waiting for us to decide on a clean way to create a more structured, "variadic" invariant, that can take a `source_locationt` and anything else we might want to decide to pass to it in the future - it's the last comment Peter made on his last review.

**Comment by tautschnig** [comment]
> Yes, this is waiting for us to decide on a clean way to create a more structured, "variadic" invariant, that can take a `source_locationt` and anything else we might want to decide to pass to it in the future - it's the last comment Peter made on his last review.

Apart from the fact that the specific bit of code that @peterschrammel commented on is now gone from the repository: can't we just solve this by doing `source_location.as_string()`? I'm sure there are more beautiful ways of doing this, but perfection is the enemy of good. And here such decision-taking might just be blocking progress?

**Comment by NlightNFotis** [comment]
@tautschnig I have now rebased the PR and I'm waiting for CI to pass. Then it should be good to go. We decided to go ahead with this now and we can implement the structured invariants later.

File: src/cbmc/symex_coverage.cpp [hunk 29]
```diff
@@ -139,15 +139,17 @@ goto_program_coverage_recordt::goto_program_coverage_recordt(
...
+  PRECONDITION(gf_it->second.body_available());
 
   // identify the file name, inlined functions aren't properly
   // accounted for
   goto_programt::const_targett end_function=
     --gf_it->second.body.instructions.end();
-  assert(end_function->is_end_function());
+  DATA_INVARIANT(
+    end_function->is_end_function(),
+    "Last instruction in a function body is end function");
```
Comment (tautschnig) [inline src/cbmc/symex_coverage.cpp]: Error messages should start with a lowercase letter. (Here and elsewhere in this PR.)

File: src/cbmc/cbmc_solvers.h [hunk 30]
```diff
@@ -68,13 +68,13 @@ class cbmc_solverst:public messaget
 
     prop_convt &prop_conv() const
     {
-      assert(prop_conv_ptr!=nullptr);
+      PRECONDITION(prop_conv_ptr!=nullptr);
```
Comment (tautschnig) [inline src/cbmc/cbmc_solvers.h]: Nit pick: clang-format.

File: src/cbmc/symex_coverage.cpp [hunk 31]
```diff
@@ -260,11 +262,15 @@ void goto_program_coverage_recordt::compute_coverage_lines(
...
+      DATA_INVARIANT(
+        c_entry->second.size() == 1 || is_branch,
+        "Instructions other than branch instructions have exactly 1 successor");
 
       for(const auto &cov : c_entry->second)
       {
-        assert(cov.second.num_executions>0);
+        DATA_INVARIANT(
+          cov.second.num_executions > 0,
+          "coverage entries can only exist with only one execution");
```
Comment (tautschnig) [inline src/cbmc/symex_coverage.cpp]: With "at least one" execution, judging by the condition that's being tested.

File: src/cbmc/cbmc_solvers.cpp [hunk 32]
```diff
@@ -164,8 +166,10 @@ std::unique_ptr<cbmc_solverst::solvert> cbmc_solverst::get_smt2(
   {
     if(solver==smt2_dect::solvert::GENERIC)
     {
-      error() << "please use --outfile" << eom;
-      throw 0;
+      throw invalid_user_input_exceptiont(
+        "required filename not provided",
+        "--outfile",
+        "provide a filename with --outfile");
```
Comment (tautschnig) [inline src/cbmc/cbmc_solvers.cpp:172]: Consistently use uppercase or lowercase for those messages (it's all uppercase in src/cbmc/bmc.cpp).

File: src/cbmc/cbmc_solvers.h [hunk 33]
```diff
@@ -68,13 +68,13 @@ class cbmc_solverst:public messaget
 
     prop_convt &prop_conv() const
     {
-      PRECONDITION(prop_conv_ptr!=nullptr);
+      PRECONDITION(prop_conv_ptr != nullptr);
```
Comment (tautschnig) [inline src/cbmc/cbmc_solvers.h]: Ah, that seems to be fixing my clang-format comment placed on the earlier commit.

File: src/util/exception_utils.cpp [hunk 34]
```diff
@@ -15,6 +15,6 @@ std::string invalid_user_input_exceptiont::what() const noexcept
   res += "Option: " + option + "\n";
   res += "Reason: " + reason;
   // Print an optional correct usage message assuming correct input parameters have been passed
-  res += correct_input.empty() ? "\n" : " Try: " + correct_input + "\n";
+  res += correct_input.empty() ? "\n" : correct_input + "\n";
```
Comment (tautschnig) [inline src/util/exception_utils.cpp]: That simplifies to `res += correct_input + "\n";` as far as I'm aware?!

File: src/cbmc/cbmc_parse_options.h [hunk 35]
```diff
@@ -76,10 +76,9 @@ class optionst;
...
+                            public xml_interfacet,
+                            public messaget
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.h]: Marked as resolved but the change is still there?

**Review by tautschnig — COMMENTED** [review]
There are still changes in the individual commits that just modify what was added in one of the other commits in this series. Some soft reset and the use of `git add -p` seems to be due.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 3bd4322721417555ec025b2e8e6b6f0cc0ca0049).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

File: src/cbmc/cbmc_solvers.cpp (hunk 32 above)
Comment (NlightNFotis) [reply src/cbmc/cbmc_solvers.cpp:172]: I changed the ones in `src/cbmc/bmc.cpp` to use lowercase, to make it more consistent with the rest of the error messages across CBMC.

**Comment by NlightNFotis** [comment]
@tautschnig I have implemented all of the changes suggested in your comments, and Peter's comments (bar the structured invariant). I have also pinged Peter for a re-review. Can you take a look at it and green-light it if you think it's good to go?

**Review by peterschrammel — APPROVED** [review]

File: src/cbmc/cbmc_solvers.h [hunk 36]
```diff
@@ -68,13 +68,13 @@ class cbmc_solverst:public messaget
...
       return *prop_conv_ptr;
     }
 
     propt &prop() const
     {
-      PRECONDITION(prop_ptr!=nullptr);
+      PRECONDITION(prop_ptr != nullptr);
```
Comment (tautschnig) [inline src/cbmc/cbmc_solvers.h]: This is still in the wrong commit, fixing up a change from an earlier commit.

File: src/cbmc/cbmc_main.cpp [hunk 37]
```diff
@@ -43,7 +43,7 @@ int main(int argc, const char **argv)
 #endif
   cbmc_parse_optionst parse_options(argc, argv);
 
-  int res=parse_options.main();
+  int res = parse_options.main();
```
Comment (tautschnig) [inline src/cbmc/cbmc_main.cpp:46]: Unnecessary change?

File: src/util/exception_utils.cpp [hunk 38]
```diff
@@ -0,0 +1,20 @@
...
+\*******************************************************************/
+
+#include "exception_utils.h"
+
+std::string invalid_user_input_exceptiont::what() const noexcept
+{
+  std::string res;
+  res += "\nInvalid User Input\n";
+  res += "Option: " + option + "\n";
+  res += "Reason: " + reason;
```
Comment (tautschnig) [inline src/util/exception_utils.cpp:16]: Wishlist: please add a test for this, i.e., intentionally use CBMC with incorrect options.

**Review by tautschnig — APPROVED** [review]

File: src/util/exception_utils.cpp (hunk 38 above)
Comment (NlightNFotis) [reply src/util/exception_utils.cpp:16]: Do you need this as part of this PR, or can it go in as part of a follow up PR for example?

Comment (tautschnig) [reply src/util/exception_utils.cpp:16]: Up to you, as long as it happens at some point :-)

Comment (chrisr-diffblue) [reply src/util/exception_utils.cpp:16]: We plan to make some further improvements to user error handling, so we'll definitely be adding suitable test cases as part of that work, so I'd be happy for this PR to go in, with tests following as part of the later improvements.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 2e6d6a0eb9106be0d8e459dccee3bffbff5af48e).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
